[DONE] /design Step 0-pre false-positive 'unexpanded template literal' abort

/design Step 0-pre aborts with false-positive "unexpanded template literal" — loader expands the guard's own RHS sentinel

## Symptom

Every `/design` invocation aborts on its **first** Bash block (Step 0-pre, public-argv validation) with:

```
/design Step 0-pre: CLAUDE_PLUGIN_ROOT is the unexpanded template literal — skill loader must expand ${CLAUDE_PLUGIN_ROOT} before Bash runs; abort
Exit code 1
```

This happens on the success path — `CLAUDE_PLUGIN_ROOT` is in fact set to a real, existing plugin directory. The skill never reaches `parse-design-argv.sh`, so `/design` is fully blocked at Step 0-pre under the affected Claude Code loader.

Observed with the cached plugin at `skills/design/SKILL.md` (larch 47.0.65).

## Offending code

`skills/design/SKILL.md`, Step 0-pre Bash fence (raw source lines 130–138):

```bash
export CLAUDE_PLUGIN_ROOT='${CLAUDE_PLUGIN_ROOT}'                       # line 130
if [ -z "${CLAUDE_PLUGIN_ROOT:-}" ]; then                              # line 131  (empty guard — robust)
  ...
fi
if [ "${CLAUDE_PLUGIN_ROOT:-}" = '${CLAUDE_PLUGIN_ROOT}' ]; then       # line 135  (unexpanded-literal guard — BROKEN)
  printf '%s\n' '/design Step 0-pre: CLAUDE_PLUGIN_ROOT is the unexpanded template literal — skill loader must expand ${CLAUDE_PLUGIN_ROOT} before Bash runs; abort' >&2
  exit 1
fi
```

The line-135 guard is the only occurrence of this `= '${CLAUDE_PLUGIN_ROOT}'` equality-against-sentinel pattern in the whole plugin (grep across `skills/` and `agents/` returns exactly one hit), so the blast radius is isolated to `/design`.

## Root cause

The line-135 guard was meant to catch the case where the skill loader (the prompt-render layer that substitutes `${CLAUDE_PLUGIN_ROOT}` → real path **before** Bash runs) *failed* to expand line 130, leaving `export CLAUDE_PLUGIN_ROOT='${CLAUDE_PLUGIN_ROOT}'` verbatim. In that failure case the exported value is the literal string `${CLAUDE_PLUGIN_ROOT}` (single-quoted, no shell expansion), and comparing it against a literal `${CLAUDE_PLUGIN_ROOT}` would detect the failure.

For that to work, the guard's **right-hand sentinel** `'${CLAUDE_PLUGIN_ROOT}'` must survive the loader as a literal token while the `export` line gets expanded. The guard therefore silently assumes a **non-uniform** loader: expands line 130 but leaves line 135's single-quoted RHS alone.

The Claude Code loader running here does a **uniform, exact-token** substitution of the bare token `${CLAUDE_PLUGIN_ROOT}` across the *entire* SKILL.md text — including the guard's own single-quoted RHS sentinel and the error string. Proof of the substitution rule, taken from the rendered prompt actually delivered to the agent:

```
RAW source line 135:    if [ "${CLAUDE_PLUGIN_ROOT:-}" = '${CLAUDE_PLUGIN_ROOT}' ]; then
RENDERED to agent:      if [ "${CLAUDE_PLUGIN_ROOT:-}" = '/Users/.../larch/47.0.65' ]; then
```

The bare `${CLAUDE_PLUGIN_ROOT}` (RHS) was replaced with the real path; `${CLAUDE_PLUGIN_ROOT:-}` (LHS — a *different* literal because of the `:-}` suffix) was left intact. So at runtime the export sets `CLAUDE_PLUGIN_ROOT=/real/path`, the LHS expands to `/real/path`, the RHS literal is `/real/path`, and the equality holds → false-positive abort.

The guard is logically incapable of discriminating success from failure under a uniform expander:

| Loader behavior | line 130 `export` | line 135 RHS | guard result |
|---|---|---|---|
| Expands everything (this env) | `/real/path` | `/real/path` | equal → **fires (WRONG — success path)** |
| Expands nothing | literal token | literal token | equal → fires (correct — failure path) |

In both consistent cases the two sides are equal, so the equality always fires. The author's mental model (expand LHS-relevant line, not the RHS sentinel) does not hold for this loader. This is a **loader-version-dependent latent bug**: it passes on loaders that don't expand single-quoted RHS bodies and 100%-false-positives on loaders that do.

## Suggested fix (primary): delete the line-135–138 guard

The guard adds no coverage that the surrounding checks don't already provide, and it is the sole source of the false positive:

- Line 131 `[ -z "${CLAUDE_PLUGIN_ROOT:-}" ]` already catches an empty expansion. (LHS uses `${CLAUDE_PLUGIN_ROOT:-}`, which the exact-token loader leaves alone, so it is a true runtime check.)
- Line 139 `[ ! -x "${CLAUDE_PLUGIN_ROOT}/skills/design/scripts/parse-design-argv.sh" ]` already catches a wrong/literal/bogus root — the parser won't be executable at a non-path.

Removing lines 135–138 eliminates the false positive with no loss of real protection. This is the minimal, surgical change.

## Suggested fix (alternative): de-tokenize the sentinel so the loader can't match it

If an explicit "verbatim-token" diagnostic is still desired, build the sentinel so its **source text** never contains the exact token `${CLAUDE_PLUGIN_ROOT}` (so the exact-token loader won't substitute it), e.g.:

```bash
_cpr_literal='$''{CLAUDE_PLUGIN_ROOT}'     # two adjacent quoted strings → literal ${CLAUDE_PLUGIN_ROOT} at bash parse time,
                                           # but the source has '$''{...} so the loader's exact-token match misses it
if [ "${CLAUDE_PLUGIN_ROOT:-}" = "$_cpr_literal" ]; then
  printf '%s\n' '/design Step 0-pre: CLAUDE_PLUGIN_ROOT is the unexpanded template literal; abort' >&2
  exit 1
fi
```

(The error string on line 136 also contains the exact token and gets expanded — cosmetic only, but worth de-tokenizing the same way if kept.)

Recommendation: prefer the primary fix (delete) for simplicity; lines 131 + 139 fully cover the real failure modes.

## Verification

- `parse-design-argv.sh 3413` runs correctly in isolation and emits the expected KVs (`POSITIONAL_KIND=issue`, `POSITIONAL_VALUE=3413`, all flags `false`), confirming the parser is fine and only the guard is at fault.
- `scripts/test-design-structure.sh` should be extended to assert that the Step 0-pre fence does **not** contain a `= '${CLAUDE_PLUGIN_ROOT}'` equality (or, if the alternative fix lands, that the sentinel source is de-tokenized), so the regression cannot reappear.

## Severity

High — blocks `/design` entirely at Step 0-pre on any Claude Code loader that uniformly expands `${CLAUDE_PLUGIN_ROOT}`. The agent had to bypass the guard and call `parse-design-argv.sh` directly to proceed.

<!-- larch:plan:start -->
## Plan


### Summary
Fix the `/design` Step 0-pre false-positive abort. De-tokenize the guard's sentinel and its diagnostic messages so a uniform-token loader cannot expand them, keep the explicit "unexpanded template literal" diagnostic, and add a regression assertion. Also de-tokenize the Step 0a twin empty-guard message.

### Files to modify/create

### UPDATED: `skills/design/SKILL.md`

Two Bash fences change. Do NOT touch the line-139 `-x` parser-executable check or its line-140 path message — they require the loader-expanded real path (hard constraint).

Step 0-pre fence — introduce a de-tokenized literal, then rewrite the sentinel and the two diagnostics.

Before:

```
export CLAUDE_PLUGIN_ROOT='${CLAUDE_PLUGIN_ROOT}'
if [ -z "${CLAUDE_PLUGIN_ROOT:-}" ]; then
  printf '%s\n' '/design Step 0-pre: CLAUDE_PLUGIN_ROOT is empty after export — skill loader must expand ${CLAUDE_PLUGIN_ROOT} in the template line before Bash runs; abort' >&2
  exit 1
fi
if [ "${CLAUDE_PLUGIN_ROOT:-}" = '${CLAUDE_PLUGIN_ROOT}' ]; then
  printf '%s\n' '/design Step 0-pre: CLAUDE_PLUGIN_ROOT is the unexpanded template literal — skill loader must expand ${CLAUDE_PLUGIN_ROOT} before Bash runs; abort' >&2
  exit 1
fi
```

After:

```
export CLAUDE_PLUGIN_ROOT='${CLAUDE_PLUGIN_ROOT}'
_cpr_literal='$''{CLAUDE_PLUGIN_ROOT}'
if [ -z "${CLAUDE_PLUGIN_ROOT:-}" ]; then
  printf '%s\n' "/design Step 0-pre: CLAUDE_PLUGIN_ROOT is empty after export — skill loader must expand ${_cpr_literal} in the template line before Bash runs; abort" >&2
  exit 1
fi
if [ "${CLAUDE_PLUGIN_ROOT:-}" = "$_cpr_literal" ]; then
  printf '%s\n' "/design Step 0-pre: CLAUDE_PLUGIN_ROOT is the unexpanded template literal ${_cpr_literal} — skill loader must expand it before Bash runs; abort" >&2
  exit 1
fi
```

Mechanism: `'$''{CLAUDE_PLUGIN_ROOT}'` is two adjacent single-quoted strings. Bash builds the literal `${CLAUDE_PLUGIN_ROOT}` at parse time, but the source text never contains the contiguous `${CLAUDE_PLUGIN_ROOT}` token, so the loader's exact-token substitution misses it. The sentinel now compares the runtime value (`${CLAUDE_PLUGIN_ROOT:-}`, which the loader leaves alone because of the `:-}` suffix) against the true literal token — the real discriminator. The two diagnostic messages switch to double quotes and reference `${_cpr_literal}`, so they print the literal token name instead of an expanded path. The substring "unexpanded template literal" stays present, preserving the existing test pin. Note: this exact-token loader behavior is reproduction-derived from the issue's RAW->RENDERED proof (the issue body / `feature-description.txt`), not a repo-normative loader contract; the fix targets that documented loader.

Step 0a fence — de-tokenize the twin empty-guard message only. Step 0a has no equality guard, so add the literal and reword the one message; do not add a new guard.

Before:

```
export CLAUDE_PLUGIN_ROOT='${CLAUDE_PLUGIN_ROOT}'
if [ -z "${CLAUDE_PLUGIN_ROOT:-}" ]; then
  printf '%s\n' '/design Step 0: CLAUDE_PLUGIN_ROOT is empty after export — skill loader must expand ${CLAUDE_PLUGIN_ROOT} in the template line before Bash runs; abort' >&2
  exit 1
fi
```

After:

```
export CLAUDE_PLUGIN_ROOT='${CLAUDE_PLUGIN_ROOT}'
_cpr_literal='$''{CLAUDE_PLUGIN_ROOT}'
if [ -z "${CLAUDE_PLUGIN_ROOT:-}" ]; then
  printf '%s\n' "/design Step 0: CLAUDE_PLUGIN_ROOT is empty after export — skill loader must expand ${_cpr_literal} in the template line before Bash runs; abort" >&2
  exit 1
fi
```

### UPDATED: `scripts/test-design-structure.sh`

Add a negative regression assertion to the Step 0-pre assertion group (beside the existing `unexpanded template literal` and `$ARGUMENTS` checks). Keep both existing positive pins unchanged. Match the existing `# shellcheck disable=SC2016` convention used nearby for grep patterns that embed `${...}`.

```
# shellcheck disable=SC2016 # Markdown literal; ${CLAUDE_PLUGIN_ROOT} must stay unexpanded in the forbidden pattern.
if printf '%s\n' "$step0pre_block" | grep -Fq "= '\${CLAUDE_PLUGIN_ROOT}'"; then
  fail 'Step 0-pre must not compare CLAUDE_PLUGIN_ROOT against a bare ${CLAUDE_PLUGIN_ROOT} sentinel (loader expands it; use a de-tokenized literal)'
fi
```

The forbidden fixed string is `= '${CLAUDE_PLUGIN_ROOT}'` with a space after `=` (the broken equality). The de-tokenized literal is written `_cpr_literal='$'...` with no space after `=`, so it does not match — the assertion fails only on a genuine regression.

### Approach
- Keep the explicit diagnostic (de-tokenize, not delete), per the operator's choice.
- Defeat the uniform-token loader by ensuring the guard's literal sentinel never appears as the contiguous `${CLAUDE_PLUGIN_ROOT}` token in source.
- Compare against the runtime export value: real path on success, literal token on loader failure.
- Leave the empty-check (`-z`) and parser-executable (`-x`) checks intact; they cover the empty-expansion and wrong-root cases.

### Edge cases
- Loader expands everything (the failing env): runtime value is the real path, not equal to the literal, so the guard passes. Bug fixed.
- Loader does not expand the seed export: runtime value equals `$_cpr_literal`, so the guard fires with a correct abort.
- Loader expands to empty (root unset at render): the `-z` check fires first.
- Wrong or bogus root: the `-x` parser check fires with the real-path message (unchanged).
- Non-exact expander (residual limitation): a future loader that also expands `${CLAUDE_PLUGIN_ROOT:-}`, matches `CLAUDE_PLUGIN_ROOT` without braces, or rewrites split-quoted fragments would defeat both the sentinel and the de-tokenize trick. The fix targets the documented exact-token loader (per the issue's reproduction); broader matching is out of scope and noted here as a known limitation.

### Failure modes
- A future edit reintroduces a bare-token sentinel. Signal: `make lint` / structure test. Mitigation: the new negative assertion fails CI.
- Switching the messages to double quotes introduces an unescaped metacharacter. Signal: render/runtime. Mitigation: the messages contain only `${_cpr_literal}` and an em-dash — no other `$`, backtick, or `"`; verified against current text.
- shellcheck flags the de-tokenize line. Signal: lint. The existing guard used the same single-quote class with no disable comment, so SKILL.md bash fences are not shellcheck-gated; the test-file assertion carries the SC2016 disable per the existing convention nearby.

### Testing strategy
- Add the negative assertion; run `bash scripts/test-design-structure.sh` (expect pass).
- Confirm the existing pins still pass: `unexpanded template literal` (test line ~271) and `parse-design-argv.sh not executable` (test line ~273) remain present in the Step 0-pre block.
- Run `bash scripts/relevant-checks.sh` (or `make lint`) after the edits.
- Manual render check: confirm the source has no contiguous `${CLAUDE_PLUGIN_ROOT}` in the sentinel or de-tokenized messages (only the intended seed export and the real-path `-x` check retain it).


## Acceptance

- `skills/design/SKILL.md` Step 0-pre no longer contains the bare-token equality `= '${CLAUDE_PLUGIN_ROOT}'`; the sentinel compares the runtime value against the de-tokenized `$_cpr_literal`.
- The two Step 0-pre "must expand" diagnostics and the Step 0a twin message reference `${_cpr_literal}` — no bare `${CLAUDE_PLUGIN_ROOT}` remains in those message strings.
- The line-139 `-x` parser-executable check and the line-140 path message are unchanged (still loader-expanded real path).
- `scripts/test-design-structure.sh` gains a negative assertion that the Step 0-pre block contains no `= '${CLAUDE_PLUGIN_ROOT}'`; the existing `unexpanded template literal` and `parse-design-argv.sh not executable` pins still pass.
- `bash scripts/test-design-structure.sh` passes, and `bash scripts/relevant-checks.sh` (or `make lint`) passes.
- A uniform-token skill loader no longer causes `/design` Step 0-pre to false-positive abort on the success path.

diff_lines: 15
<!-- larch:plan:end -->

--- a/scripts/test-design-structure.sh
+++ b/scripts/test-design-structure.sh
@@ -272,6 +272,10 @@ printf '%s\n' "$step0pre_block" | grep -Fq 'unexpanded template literal' \
   || fail 'Step 0-pre must reject unexpanded CLAUDE_PLUGIN_ROOT template literal'
 printf '%s\n' "$step0pre_block" | grep -Fq 'parse-design-argv.sh not executable' \
   || fail 'Step 0-pre must verify parse-design-argv.sh is executable before invoke'
+# shellcheck disable=SC2016 # Markdown literal; ${CLAUDE_PLUGIN_ROOT} must stay unexpanded in the forbidden pattern.
+if printf '%s\n' "$step0pre_block" | grep -Fq "= '\${CLAUDE_PLUGIN_ROOT}'"; then
+  fail 'Step 0-pre must not compare CLAUDE_PLUGIN_ROOT against a bare ${CLAUDE_PLUGIN_ROOT} sentinel (loader expands it; use a de-tokenized literal)'
+fi
 contains "$PARSE_DESIGN_ARGV_SH" 'assert_safe_kv_value' 'parse-design-argv.sh missing newline guard on emitted values'
 
 DESIGN_DRIVER_SH="$REPO_ROOT/skills/design/scripts/design-driver.sh"

--- a/skills/design/SKILL.md
+++ b/skills/design/SKILL.md
@@ -128,12 +128,13 @@ Run `parse-design-argv.sh` as the single authoritative Step 0-pre parser. Render
 
 ```bash
 export CLAUDE_PLUGIN_ROOT='${CLAUDE_PLUGIN_ROOT}'
+_cpr_literal='$''{CLAUDE_PLUGIN_ROOT}'
 if [ -z "${CLAUDE_PLUGIN_ROOT:-}" ]; then
-  printf '%s\n' '/design Step 0-pre: CLAUDE_PLUGIN_ROOT is empty after export — skill loader must expand ${CLAUDE_PLUGIN_ROOT} in the template line before Bash runs; abort' >&2
+  printf '%s\n' "/design Step 0-pre: CLAUDE_PLUGIN_ROOT is empty after export — skill loader must expand ${_cpr_literal} in the template line before Bash runs; abort" >&2
   exit 1
 fi
-if [ "${CLAUDE_PLUGIN_ROOT:-}" = '${CLAUDE_PLUGIN_ROOT}' ]; then
-  printf '%s\n' '/design Step 0-pre: CLAUDE_PLUGIN_ROOT is the unexpanded template literal — skill loader must expand ${CLAUDE_PLUGIN_ROOT} before Bash runs; abort' >&2
+if [ "${CLAUDE_PLUGIN_ROOT:-}" = "$_cpr_literal" ]; then
+  printf '%s\n' "/design Step 0-pre: CLAUDE_PLUGIN_ROOT is the unexpanded template literal ${_cpr_literal} — skill loader must expand it before Bash runs; abort" >&2
   exit 1
 fi
 if [ ! -x "${CLAUDE_PLUGIN_ROOT}/skills/design/scripts/parse-design-argv.sh" ]; then
@@ -245,8 +246,9 @@ On success, Step 0b consumes the bound mental booleans, optional `run_id`, `POSI
 
 ```bash
 export CLAUDE_PLUGIN_ROOT='${CLAUDE_PLUGIN_ROOT}'
+_cpr_literal='$''{CLAUDE_PLUGIN_ROOT}'
 if [ -z "${CLAUDE_PLUGIN_ROOT:-}" ]; then
-  printf '%s\n' '/design Step 0: CLAUDE_PLUGIN_ROOT is empty after export — skill loader must expand ${CLAUDE_PLUGIN_ROOT} in the template line before Bash runs; abort' >&2
+  printf '%s\n' "/design Step 0: CLAUDE_PLUGIN_ROOT is empty after export — skill loader must expand ${_cpr_literal} in the template line before Bash runs; abort" >&2
   exit 1
 fi
 LARCH_TIMING_SKILL=design "${CLAUDE_PLUGIN_ROOT}/scripts/timing-ledger.sh" mark "design Step 0 — session setup" || true

